Flake8

dmentipl opened the pull request:
Addresses #1462 

This small PR adds [flake8](https://flake8.pycqa.org/) linting to GitHub Actions. This adds flake8 as a "development dependency" to the Python 3.7 conda environment file. I have also added a config block for flake8 in setup.cfg to make it compatible with black, see [here](https://black.readthedocs.io/en/stable/guides/using_black_with_other_tools.html#flake8).

Currently the action is failing to pass. There are a few changes required, some of which may require some discussion of the approach required. E.g. stylistic things such as "E731 do not assign a lambda expression, ...".

Testing:
 - [x] Ran tests and they passed OK

tjtg commented on setup.cfg:
```diff
@@ -59,6 +59,11 @@ full =
+[flake8]
+max-line-length = 88
+select = C,E,F,W
+extend-ignore = E203,W503
```
I think the issues raised by flake8 are generally good and useful. I'd be happy if someone sat down, worked through them and submitted a later "fix flake8 issues" PR that made flake8 not fail. It's probably best for one person to just do something sensible rather than discussing each item one by one.

The only flake8 issue which sticks out to me as problematic is "E731 do not assign a lambda expression, use a def". When setting up a constraint for an Iris cube, it's idiomatic to use a lambda for more complex constraints. For example, see the orange warning box at https://scitools-iris.readthedocs.io/en/stable/userguide/subsetting_a_cube.html#cube-extraction
I'd suggest adding E731 to the extend-ignore config line.

benfitzpatrick replied:
Agreed on E731, happy to ignore that one. I'll bag the fix-flake8-issues if that's OK.

dmentipl replied:
> submitted a later "fix flake8 issues" PR

Sounds good to me.

> I'd suggest adding E731 to the extend-ignore config line.

E731 refers to assigning a lambda to a variable. The example linked to passes a lambda in as a function argument which E731 is fine with. Having said that, I'm happy to ignore E731.

> I'll bag the fix-flake8-issues if that's OK.

Thanks @benfitzpatrick.

tjtg commented on .github/workflows/tests.yml:
```diff
@@ -87,6 +87,11 @@ jobs:
+      run: |
+        source '/usr/share/miniconda/etc/profile.d/conda.sh'
+        conda activate improver_${{ matrix.env }}
+        flake8 improver improver_tests
```
Suggest adding `|| true` so that github actions runs flake8, but doesn't fail the CI job yet.
Once the issues identified by flake8 are fixed (in a separate PR) that bypass can be removed and enforcement can start.

dmentipl replied:
Done.

tjtg reviewed (commented):
I think this is good.

Personally, I'd prefer this approach to add flake8 in github actions, rather than re-enable Hound. Hound used to create a lot of comment spam on PRs. This would count as a failed CI task with associated viewable logs, similar to the other code checker tools that are already run - a nicer way to report than many PR comments.

benfitzpatrick commented on setup.cfg:
```diff
@@ -59,6 +59,11 @@ full =
     timezonefinder
 
+[flake8]
+max-line-length = 88
```
I think this makes sense

dmentipl replied:
I am inclined to just fix the strings and docstrings. Although, as Emerson said, "A foolish consistency is the hobgoblin of little minds" I can set it to 100.

benfitzpatrick commented on .github/workflows/tests.yml:
```diff
@@ -87,6 +87,11 @@ jobs:
+        source '/usr/share/miniconda/etc/profile.d/conda.sh'
+        conda activate improver_${{ matrix.env }}
+        flake8 improver improver_tests
     - name: pylint
```
I think we can remove pylint now

benfitzpatrick replied:
Could we also replace the pylint command in bin/improver-tests in favour of flake8?

dmentipl replied:
Good idea. I'll do this.

There's a few files which have `# pylint: disable=...` annotations. Perhaps you could remove these in #1466?

dmentipl commented:
Thanks for the reviews @benfitzpatrick and @tjtg. I've addressed all comments:

- changed the line limit to 100
- ignored E731
- temporarily disabled the CI from failing on flake8 errors (using `|| true` as suggested)
- removed pylint

tjtg reviewed (approved):
I think all the comments have been addressed here and this is good to go now. We should try to get #1466 merged soon to resolve the issues raised by flake8.
